src/test/java/org/influxdb/dto/BoundParameterQueryTest.java

```diff
@@ -0,0 +1,188 @@
+    Assert.assertEquals(2, params.size());
+    Assert.assertEquals(params.get("abc"), 0.0);
+    Assert.assertEquals(params.get("bcd"), 10.0);
+  }
+
+  @Test
+  public void testDifferentTypePlaceHolderParsing() throws IOException {
+    BoundParameterQuery query = QueryBuilder.newQuery("SELECT * FROM abc WHERE number > $number"
+        + " AND bool = $bool AND string = $string AND other = $object")
+        .forDatabase("foobar")
+        .bind("number", 0)
+        .bind("bool", true)
+        .bind("string", "test")
+        .bind("object", new Object())
+        .create();
+    Map<String, Object> params = readObject(decode(query.getParameterJsonWithUrlEncoded()));
+    Assert.assertEquals(4, params.size());
+    Assert.assertEquals(params.get("number"), 0.0);
+    Assert.assertEquals(params.get("bool"), true);
+    Assert.assertEquals(params.get("string"), "test");
+    Assert.assertTrue(((String)params.get("object")).matches("java.lang.Object@[a-z0-9]+"));
+  }
+
+  @Test
+  public void testIgnoreInvalidPlaceholders() throws UnsupportedEncodingException {
+    BoundParameterQuery query = QueryBuilder.newQuery("SELECT * FROM abc WHERE a > $")
+        .forDatabase("foobar")
+        .create();
+    Assert.assertEquals(decode(query.getParameterJsonWithUrlEncoded()), "{}");
+
+    query = QueryBuilder.newQuery("SELECT * FROM abc WHERE a > $abc$cde").forDatabase("foobar").create();
+    Assert.assertEquals(decode(query.getParameterJsonWithUrlEncoded()), "{}");
+  }
+
+  @Test
+  public void testUnbalancedQuery() throws UnsupportedEncodingException {
+    // too many placeholders
+    try {
+      BoundParameterQuery query = QueryBuilder.newQuery("SELECT * FROM abc WHERE a > $abc AND b < $bcd")
+          .forDatabase("foobar")
+          .bind("abc", 0)
+          .create();
+      query.getParameterJsonWithUrlEncoded();
+      Assert.fail("Expected RuntimeException because of unbalanced placeholders and parameters");
+    } catch (RuntimeException rte) {
+      // expected
+    }
+
+    // too many parameters
+    try {
+      BoundParameterQuery query = QueryBuilder.newQuery("SELECT * FROM abc WHERE a > $abc")
+          .forDatabase("foobar")
+          .bind("abc", 0)
+          .bind("bcd", 10)
+          .create();
+      query.getParameterJsonWithUrlEncoded();
+      Assert.fail("Expected RuntimeException because of unbalanced placeholders and parameters");
+    } catch (RuntimeException rte) {
+      // expected
+    }
+
+    // unbound placeholder
+    try {
+      BoundParameterQuery query = QueryBuilder.newQuery("SELECT * FROM abc WHERE a > $abc")
+          .forDatabase("foobar")
+          .bind("bcd", 10)
+          .create();
+      query.getParameterJsonWithUrlEncoded();
+      Assert.fail("Expected RuntimeException because of unbound placeholder");
+    } catch (RuntimeException rte) {
+      // expected
+    }
+  }
+
+  @Test
+  public void testEqualsAndHashCode() {
```

Comment:
Is this really necessary? Did you create your own hashcode/equals implementation or you are using the one provided by your IDE?

Reply:
There is a test for hashcode and equals in Query. So I just did it for BoundParameterQuery, too.